### PR TITLE
[ValueTracking] Handle range attributes

### PR DESCRIPTION
**Pull request by andjo403**
Handle the range attribute in ValueTracking.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -1467,14 +1467,21 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
     if (MDNode *MD =
             Q.IIQ.getMetadata(cast<Instruction>(I), LLVMContext::MD_range))
       computeKnownBitsFromRangeMetadata(*MD, Known);
-    if (const Value *RV = cast<CallBase>(I)->getReturnedArgOperand()) {
+
+    const CallBase *CB = cast<CallBase>(I);
+
+    const Attribute RangeAttr = CB->getRetAttr(llvm::Attribute::Range);
+    if (RangeAttr.isValid())
+      Known = RangeAttr.getRange().toKnownBits();
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:1482]:
```suggestion
    const Attribute RangeAttr = CB->getRetAttr(llvm::Attribute::Range);
    if (std::optional<ConstantRange> Range = CB->getRange())
      Known = Range->toKnownBits();
```
I'd suggest adding a small wrapper for this, given how often we're going to repeat it...

File: llvm/unittests/Analysis/ValueTrackingTest.cpp [hunk 2]
```diff
@@ -2085,6 +2085,76 @@ TEST_F(ValueTrackingTest, KnownNonZeroFromDomCond2) {
   EXPECT_EQ(isKnownNonZero(A, DL, 0, &AC, CxtI2, &DT), false);
 }
 
+TEST_F(ValueTrackingTest, KnownNonZeroFromRangeAttributeArgument) {
```
Comment (nikic) [inline llvm/unittests/Analysis/ValueTrackingTest.cpp:2088]: Please use lit tests instead (usually via llvm/test/Transforms/InstSimplify).

**Comment by andjo403** [comment]
@nikic  fixed the comments and split up in multiple commits to get the pre commit test looking good.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -1467,14 +1467,20 @@ static void computeKnownBitsFromOperator(const Operator *I,
...
     // function.
     if (MDNode *MD =
             Q.IIQ.getMetadata(cast<Instruction>(I), LLVMContext::MD_range))
       computeKnownBitsFromRangeMetadata(*MD, Known);
-    if (const Value *RV = cast<CallBase>(I)->getReturnedArgOperand()) {
+
+    const CallBase *CB = cast<CallBase>(I);
+
+    if (std::optional<ConstantRange> Range = CB->getRange())
+      Known = Range->toKnownBits();
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:1481]: Think this should be `Known.unionWith(Range->toKnownBits())`.

Comment (andjo403) [reply llvm/lib/Analysis/ValueTracking.cpp:1481]: hmm maybe it was suggested to forbid range metadata on calls in the future https://github.com/llvm/llvm-project/pull/84627#discussion_r1518589822 but maybe needed until then or at least swap the order so this is done before the range metadata handling

Comment (andjo403) [reply llvm/lib/Analysis/ValueTracking.cpp]: fixed now

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 4]
```diff
@@ -2893,6 +2904,18 @@ bool isKnownNonZero(const Value *V, const APInt &DemandedElts, unsigned Depth,
     }
   }
 
+  std::optional<ConstantRange> Range;
+  if (const CallBase *CB = dyn_cast<CallBase>(V))
+    Range = CB->getRange();
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2909]: Could you please rebase over https://github.com/llvm/llvm-project/commit/2cc75aed0925f3c7e8e50b2f4bdc94daac4cb08d and move this part into the CallBase handling code in isKnownNonZeroFromOperator()?

**Comment by nikic** [comment]
> the lit tests feels a bit fragile as multiple of the functions is called in many of the cases eg for the icmp test if the isKnownNonZero and computeConstantRange is removed the test will still pass du e to the ComputeKnownBits

If you test `x_with_range | y_unknown == 0`, then that's a way to check only isKnownNonZero() as opposed to the others.

**Comment by andjo403** [comment]
rebased and fixed the comments. good tip about the isKnownNonZero tests.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/include/llvm/IR/InstrTypes.h [hunk 5]
```diff
@@ -2154,6 +2155,10 @@ class CallBase : public Instruction {
   /// parameter.
   FPClassTest getParamNoFPClass(unsigned i) const;
 
+  /// If this return value has a range attribute, return the value range of the
+  /// argument. Otherwise, std::nullopt is returned.
```
Comment (dtcxzyw) [inline llvm/include/llvm/IR/InstrTypes.h:2159]: typo

**Pull request merged**